sdk/servicebus/service-bus/src/serviceBusMessageBatch.ts

```diff
@@ -0,0 +1,190 @@
+    this._sizeInBytes = 0;
+    this._count = 0;
+  }
+
+  /**
+   * @property The maximum size of the batch, in bytes.
+   * @readonly
+   */
+  get maxSizeInBytes(): number {
+    return this._maxSizeInBytes;
+  }
+
+  /**
+   * @property Size of the `ServiceBusMessageBatch` instance after the messages added to it have been
+   * encoded into a single AMQP message.
+   * @readonly
+   */
+  get sizeInBytes(): number {
+    return this._sizeInBytes;
+  }
+
+  /**
+   * @property Number of messages in the `ServiceBusMessageBatch` instance.
+   * @readonly
+   */
+  get count(): number {
+    return this._count;
+  }
+
+  /**
+   * @property Represents the single AMQP message which is the result of encoding all the events
+   * added into the `ServiceBusMessageBatch` instance.
+   *
+   * This is not meant for the user to use directly.
+   *
+   * When the `ServiceBusMessageBatch` instance is passed to the `sendBatch()` method on the `Sender`,
+   * this single batched AMQP message is what gets sent over the wire to the service.
+   * @readonly
+   */
+  get _message(): Buffer | undefined {
+    return this._batchMessage;
+  }
+
+  /**
+   * Tries to add a message to the batch if permitted by the batch's size limit.
+   * **NOTE**: Always remember to check the return value of this method, before calling it again
+   * for the next message.
+   *
+   * @param message  An individual service bus message.
+   * @returns A boolean value indicating if the message has been added to the batch or not.
+   */
+  public tryAdd(message: ServiceBusMessage): boolean {
+    throwTypeErrorIfParameterMissing(this._context.namespace.connectionId, "tryAdd", "message");
+
+    // Convert ServiceBusMessage to AmqpMessage.
+    const amqpMessage = toAmqpMessage(message);
+    amqpMessage.body = this._context.namespace.dataTransformer.encode(message.body);
+
+    // Encode every amqp message and then convert every encoded message to amqp data section
+    this._encodedMessages.push(RheaMessageUtil.encode(amqpMessage));
+
+    const batchMessage: AmqpMessage = {
+      body: RheaMessageUtil.data_sections(this._encodedMessages)
+    };
+
+    if (amqpMessage.message_annotations) {
+      batchMessage.message_annotations = amqpMessage.message_annotations;
+    }
+    if (amqpMessage.application_properties) {
+      batchMessage.application_properties = amqpMessage.application_properties;
+    }
+    for (const prop of messageProperties) {
+      if ((amqpMessage as any)[prop]) {
+        (batchMessage as any)[prop] = (amqpMessage as any)[prop];
+      }
+    }
+
+    const encodedBatchMessage = RheaMessageUtil.encode(batchMessage);
+    const currentSize = encodedBatchMessage.length;
+
+    // this._batchMessage will be used for final send operation
+    if (currentSize > this._maxSizeInBytes) {
+      this._encodedMessages.pop();
+      return false;
+    }
+    this._batchMessage = encodedBatchMessage;
+    this._sizeInBytes = currentSize;
+    this._count++;
```

Comment:
Should this be the same as _encodedMessages.length? Ie, can we remove this field and just use the array length?

Reply:
Good idea, I'll remove.